___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" 
alt="CLRSWY"></p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#9d4f8c; font-size:100%; text-align:center; border-radius:10px 10px;">WAY TO REINVENT YOURSELF</p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#060108; font-size:200%; text-align:center; border-radius:10px 10px;">Data Analysis with Python</p>

![Image_Assignment](https://i.ibb.co/cY9HhF7/bnote-assignment.gif)

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#4d77cf; font-size:200%; text-align:center; border-radius:10px 10px;">Assignment 02 (Pandas)</p>

<a id="toc"></a>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">CONTENT</p>


* [CONTEXT](#0)
* [ABOUT THE DATASET & ATRIBUTES](#00)
* [INSTRUCTIONS (Tasks in the Assignment)](#000)
* [IMPORTING LIBRARIES NEEDED IN THIS NOTEBOOK](#1)
* [Task 01. Checking the head of the DataFrame](#2)
* [Task 02. How many rows and columns are there?](#3)
* [Task 03. What is the average Purchase Price?](#4)
* [Task 04. What were the highest and lowest purchase prices?](#5) 
* [Task 05. Listing the country names and number of records belongs to each country](#6)
* [Task 06. How many records are as Country United Kingdom?](#7)
* [Task 07. Creating a column named Total which shows the total amount of spend using UnitPrice and Quantity](#8)
* [Task 08. Showing total spending amount according to the Countries](#9)
* [Task 09. Which Country spend least?](#10)
* [Task 10. Calculating average of spending according to country with apply function](#11) 
* [Task 11. Calculating average, mean, max of spending according to country with agg function](#12)
* [Task 12. What is the max Total spending in whole dataset and show the records of that spending?](#13)
* [Task 13. How many people who live in Spain spend more than average Total?](#14)
* [Task 14. Converting datatype of InvoiceDate to datetime and assign it to new column name InvoiceDate2](#15)
* [Task 15. Taking only date values (year-month-day) from InvoiceDate2 and assign it to a new column named "date"](#16)
* [Task 16. Dropping InvoiceDate2 and InvoiceDate columns permanently](#17)
* [Task 17. Removing A, B, G and E characters at the first 5 rows of "StockCode" in ecom dataframe with a string method](#18) 
* [Task 18. Taking only Total and date columns with loc and craete a new df named ecom2](#19)
* [Task 19. Taking only InvoiceNo and CustomerID columns with iloc and craete a new df named ecom3](#20)
* [Task 20. Combining ecom2 and ecom3 with concat() method](#21)
* [Task 21. Combining the common (intersected) values of ecom2 and ecom3 with merge() method](#22)
* [Task 22. Combining ecom2 and ecom3 with join() method](#23)
* [THE END OF THE ASSIGNMENT 02](#24)

<a id="toc"></a>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">CONTEXT</p>

<a id="0"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

E-Commerce is currently one of the fastest and dynamically evolving industries in the world. Its popularity has been growing rapidly with the ease of digital transactions and quick door-to-door deliveries. A major source of big tech companies revenues comes from the interaction of their underlying proprietary algorithms that are heavily powered by data science, so it’s fundamental to understand the methods applied to maintain and grow the number of clients.

"This is a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers."

<a id="toc"></a>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">ABOUT THE DATASET & ATRIBUTES</p>

<a id="00"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

Typically e-commerce datasets are proprietary and consequently hard to find among publicly available data. However, The [UCI Machine Learning Repository](http://archive.ics.uci.edu/ml/index.php) has made this dataset containing actual transactions from 2010 and 2011. The dataset is maintained on their site, where it can be found by the title "Online Retail".

**Attribute Information:**

**``InvoiceNo:``** Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.<br>
**``StockCode:``** Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.<br>
**``Description:``** Product (item) name. Nominal.<br>
**``Quantity:``** The quantities of each product (item) per transaction. Numeric.<br>
**``InvoiceDate:``** Invice Date and time. Numeric, the day and time when each transaction was generated.<br>
**``UnitPrice:``** Unit price. Numeric, Product price per unit in sterling.<br>
**``CustomerID:``** Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.<br>
**``Country:``** Country name. Nominal, the name of the country where each customer resides.<br>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">INSTRUCTIONS (Tasks in the Assignment)</p>

<a id="000"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

**In this assignment you're required to handle with the following tasks:** 

- **Task 01.** Checking the head of the DataFrame.<br>
- **Task 02.** How many rows and columns are there?<br>
- **Task 03.** What is the average of Product sold and the average price of Product per unit?<br>
- **Task 04.** What were the highest and lowest purchase prices?<br>
- **Task 05.** Listing the country names and number of their records belongs to each country.<br>
- **Task 06.** How many records are as Country United Kingdom?<br>
- **Task 07.** Creating a column named "Total" which shows the total amount of spend using UnitPrice and Quantity.<br>
- **Task 08.** Showing total spending amount according to the Countries and their percentages.<br>
- **Task 09.** Which Country spend least?<br>
- **Task 10.** Calculating average of spending according to country with apply function  and create a new feature named "Average_Spend".<br>
- **Task 11.** Calculating average, min, max of spending according to country with agg function and create new features named "Max_Spend" and "Min_Spend".<br>
- **Task 12.** What is the max Total spending in whole dataset and show all records of that spending?<br>
- **Task 13.** How many people who live in Spain spend more than average Total?<br>
- **Task 14.** Converting datatype of InvoiceDate to datetime and assign it to new column name InvoiceDate2.<br>
- **Task 15.** Taking only date values (year-month-day) from InvoiceDate2 and assign it to a new column named "date".<br>
- **Task 16.** Dropping InvoiceDate2 and InvoiceDate columns permanently.<br>
- **Task 17.** Removing A, B, G and E characters at the first 5 rows of "StockCode" in ecom dataframe with a string method.<br>
- **Task 18.** Taking only Total and date columns with loc and craete a new df named ecom2.<br>
- **Task 19.** Taking only InvoiceNo and CustomerID columns with iloc and craete a new df named ecom3.<br>
- **Task 20.** Combining ecom2 and ecom3 with concat() method.<br>
- **Task 21.** Combining the common (intersected) values of ecom2 and ecom3 with merge() method.<br>
- **Task 22.** Combining ecom2 and ecom3 with join() method.<br>

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">IMPORTING LIBRARIES NEEDED IN THIS STUDY</p>

<a id="1"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

Once you've installed NumPy and Pandas you can import them as a library. Then read the dataset and assig it as **``"ecom"``**:

In [2]:
# Import the related library to this task

# YOUR CODE IS HERE 
import numpy as np
import pandas as pd
import seaborn as sns

import matplotlib.pyplot as plt

In [3]:
# After reading the dataset, assign it named "ecom"

# YOUR CODE IS HERE 
ecom = pd.read_csv("data.csv", encoding = "ISO-8859-1")
ecom

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 01: Check the Head of the DataFrame</p>

<a id="2"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [4]:
# YOUR CODE IS HERE 

ecom.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7  12/1/2010 8:28       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

Desired Output:

![image.png](https://i.ibb.co/LdKJHb6/assignment1a.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 02: How many rows and columns are there?</p>

<a id="3"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [5]:
# YOUR CODE IS HERE 

ecom.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [6]:
# YOUR CODE IS HERE 

(len(ecom.index), len(ecom.columns))

(541909, 8)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 03: What is the average of Product sold and the average price of Product per unit?</p>

<a id="4"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [7]:
# YOUR CODE IS HERE 

a = ecom["UnitPrice"].mean()
a

4.611113626088513

In [8]:
# YOUR CODE IS HERE 

b = ecom["Quantity"].mean()
b

9.55224954743324

In [9]:
# YOUR CODE IS HERE 

pd.Series(data = (a, b), index = ["UnitPrice", "Quantity"])


UnitPrice    4.611114
Quantity     9.552250
dtype: float64

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 04: What are the highest and lowest purchase prices?</p>

<a id="5"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [10]:
# YOUR CODE IS HERE 

ecom["UnitPrice"].max()


38970.0

In [11]:
# YOUR CODE IS HERE 

ecom["UnitPrice"].min()


-11062.06

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 05: List the country names and number of their records belongs to each country</p>

<a id="6"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [12]:
ecom

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [13]:
# YOUR CODE IS HERE 

ecom["Country"].value_counts()



United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58
Lebanon 

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 06: How many records are there in the United Kingdom?</p>

<a id="7"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [14]:
# YOUR CODE IS HERE 

ecom[ecom["Country"] == "United Kingdom"].count()

InvoiceNo      495478
StockCode      495478
Description    494024
Quantity       495478
InvoiceDate    495478
UnitPrice      495478
CustomerID     361878
Country        495478
dtype: int64

### Desired Output:

InvoiceNo      495478
StockCode      495478
Description    494024
Quantity       495478
InvoiceDate    495478
UnitPrice      495478
CustomerID     361878
Country        495478
dtype: int64

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 07: Create a column named "Total" which shows the total amount of spend using UnitPrice and Quantity</p>

<a id="8"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [15]:
# YOUR CODE IS HERE 

ecom


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [16]:
total = ecom["UnitPrice"] * ecom["Quantity"]
total

0         15.30
1         20.34
2         22.00
3         20.34
4         20.34
          ...  
541904    10.20
541905    12.60
541906    16.60
541907    16.60
541908    14.85
Length: 541909, dtype: float64

In [17]:
# YOUR CODE IS HERE 

ecom["total"] = ecom["UnitPrice"] * ecom["Quantity"]
ecom

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  total  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  15.30  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  22.00  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
...                 ...        ...         ...             ...    ...  
541904  12/9/2011 12:50       0.85     12680.0          France  10.20  
541905  12/9/2011 12:50       2.10     12680.0          France  12.60  
541906  12/9/2011 12:50       4.15     12680.0          France  16.60  
541907  12/9/2011 12:50       4.15     12680.0          France  16.60  
541908  12/9/2011 12:50       4.95     12680.0          France  14.85  

[541909 rows x 9 columns]

In [18]:
ecom.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  total  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  22.00  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34

Desired Output:

![image.png](https://i.ibb.co/7y5d0sp/assignment2b.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 08: Show total spending amount according to the Countries and their percentages.</p>

<a id="9"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [19]:
# YOUR CODE IS HERE 

ecom

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  total  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  15.30  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  22.00  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
...                 ...        ...         ...             ...    ...  
541904  12/9/2011 12:50       0.85     12680.0          France  10.20  
541905  12/9/2011 12:50       2.10     12680.0          France  12.60  
541906  12/9/2011 12:50       4.15     12680.0          France  16.60  
541907  12/9/2011 12:50       4.15     12680.0          France  16.60  
541908  12/9/2011 12:50       4.95     12680.0          France  14.85  

[541909 rows x 9 columns]

In [20]:
ecom.groupby("Country")["total"].sum()

Country
Australia                137077.270
Austria                   10154.320
Bahrain                     548.400
Belgium                   40910.960
Brazil                     1143.600
Canada                     3666.380
Channel Islands           20086.290
Cyprus                    12946.290
Czech Republic              707.720
Denmark                   18768.140
EIRE                     263276.820
European Community         1291.750
Finland                   22326.740
France                   197403.900
Germany                  221698.210
Greece                     4710.520
Hong Kong                 10117.040
Iceland                    4310.000
Israel                     7907.820
Italy                     16890.510
Japan                     35340.620
Lebanon                    1693.880
Lithuania                  1661.060
Malta                      2505.470
Netherlands              284661.540
Norway                    35163.460
Poland                     7213.140
Portugal            

In [22]:
# YOUR CODE IS HERE 

ecom.groupby("Country")["total"].sum().sort_values(ascending = False) / ecom["Total"].sum() * 100

KeyError: 'Total'

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 09: Which Country spend least?</p>

<a id="10"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 




### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 10: Calculate the average of spending according to country with apply function and create a new feature named "Average_Spend"</p>

<a id="11"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/b3m1jhs/assignment2c.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 11: Calculate average, min, max of spending according to country with agg() function and create new features named "Max_Spend" and "Min_Spend"</p>

<a id="12"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/0qRCRjH/assignment2d.png)

In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/Smd3Tkt/assignment2e.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 12: What is the max Total spending in whole dataset and show all records of that spending?</p>

<a id="13"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/2W9t4nh/assignment2f.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 13: How many people who live in Spain spend more than average Total?</p>

<a id="14"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 14: Convert datatype of "InvoiceDate" to datetime and assign it to new column name "InvoiceDate2"</p>

<a id="15"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



In [ ]:
# YOUR CODE IS HERE 



### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 15: Take only date values (year-month-day) from InvoiceDate2 and assign it to a new column named **``date``**</p>

<a id="16"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/hdQm0jx/assignment2g.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 16: Drop "InvoiceDate2" and "InvoiceDate" columns permanently</p>

<a id="17"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/2dPtL0k/assignment2h.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 17: Remove A, B, G and E characters at the first 5 rows of "StockCode" in ecom dataframe with a string method</p>

<a id="18"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 18: Take only "Total" and "date" columns with loc and craete a new df named **``ecom2``**</p>

<a id="19"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/yFyxdCz/assignment2j.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 19: Take only "InvoiceNo" and "CustomerID" columns with iloc and craete a new df named **``ecom32``**</p>

<a id="20"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/9HsWScR/assignment2k.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 20: Combine "ecom2" and "ecom3" with concat() method</p>

<a id="21"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/6NGsVMy/assignment2l.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 21: Combine the common (intersected) values of "ecom2" and "ecom3" with merge() method</p>

<a id="22"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

**If you "MemoryError" occurs, take only the first 100 rows from bot DFs**

In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/3cgVkjv/assignment2m.png)

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Task 22: Combine "ecom2" and "ecom3" with join() method</p>

<a id="23"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

In [ ]:
# YOUR CODE IS HERE 



Desired Output:

![image.png](https://i.ibb.co/DfS2KqQ/assignment2n.png)

## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">THE END OF THE ASSIGNMENT 02</p>

# Now. it's time to relax!

![Image_Assignment](https://i.ibb.co/jZD3Mqb/relaxing.jpg)

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" 
alt="CLRSWY"></p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#9d4f8c; font-size:100%; text-align:center; border-radius:10px 10px;">WAY TO REINVENT YOURSELF</p>

<a id="24"></a>
<a href="#toc" class="btn btn-primary btn-sm" role="button" aria-pressed="true" 
style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>
___